lnwallet/chanfunding/psbt_assembler.go

```diff
@@ -0,0 +1,490 @@
+// to construct a funding output and channel point. An optional base PSBT can
+// be supplied which will be used to add the channel output to instead of
+// creating a new one.
+func NewPsbtAssembler(fundingAmt btcutil.Amount, basePsbt *psbt.Packet,
+	netParams *chaincfg.Params) *PsbtAssembler {
+
+	return &PsbtAssembler{
+		fundingAmt: fundingAmt,
+		basePsbt:   basePsbt,
+		netParams:  netParams,
+	}
+}
+
+// ProvisionChannel creates a new ShimIntent given the passed funding Request.
+// The returned intent is immediately able to provide the channel point and
+// funding output as they've already been created outside lnd.
+//
+// NOTE: This method satisfies the chanfunding.Assembler interface.
+func (p *PsbtAssembler) ProvisionChannel(req *Request) (Intent, error) {
+	// We'll exit out if this field is set as the funding transaction will
+	// be assembled externally, so we don't influence coin selection.
+	if req.SubtractFees {
+		return nil, fmt.Errorf("SubtractFees not supported for PSBT")
+	}
+
+	intent := &PsbtIntent{
+		ShimIntent: ShimIntent{
+			localFundingAmt: p.fundingAmt,
+		},
+		State:     PsbtShimRegistered,
+		BasePsbt:  p.basePsbt,
+		PsbtReady: make(chan error, 1),
+		netParams: p.netParams,
+	}
+
+	// A simple sanity check to ensure the provisioned request matches the
+	// re-made shim intent.
+	if req.LocalAmt+req.RemoteAmt != p.fundingAmt {
+		return nil, fmt.Errorf("intent doesn't match PSBT "+
+			"assembler: local_amt=%v, remote_amt=%v, funding_amt=%v",
+			req.LocalAmt, req.RemoteAmt, p.fundingAmt)
+	}
+
+	return intent, nil
+}
+
+// FundingTxAvailable is an empty method that an assembler can implement to
+// signal to callers that its able to provide the funding transaction for the
+// channel via the intent it returns.
+//
+// NOTE: This method is a part of the FundingTxAssembler interface.
+func (p *PsbtAssembler) FundingTxAvailable() {}
+
+// A compile-time assertion to ensure PsbtAssembler meets the Assembler
+// interface.
+var _ Assembler = (*PsbtAssembler)(nil)
+
+// sumUtxoInputValues tries to extract the sum of all inputs specified in the
+// UTXO fields of the PSBT. An error is returned if an input is specified that
+// does not contain any UTXO information.
+func sumUtxoInputValues(packet *psbt.Packet) (int64, error) {
+	// We take the TX ins of the unsigned TX as the truth for how many
+	// inputs there should be, as the fields in the extra data part of the
+	// PSBT can be empty.
+	if len(packet.UnsignedTx.TxIn) != len(packet.Inputs) {
+		return 0, fmt.Errorf("TX input length doesn't match PSBT " +
+			"input length")
+	}
+	inputSum := int64(0)
+	for idx, in := range packet.Inputs {
+		switch {
+		case in.WitnessUtxo != nil:
+			// Witness UTXOs only need to reference the TxOut.
+			inputSum += in.WitnessUtxo.Value
+
+		case in.NonWitnessUtxo != nil:
+			// Non-witness UTXOs reference to the whole transaction
+			// the UTXO resides in.
+			utxOuts := in.NonWitnessUtxo.TxOut
```

Comment:
Nice, I totally forgot that the inputs in a PSBT include the value as well, super convenient for verification checks like this.